### PR TITLE
[autoscaler][aws] Update _configure_key_pair to available_node_types

### LINKED ISSUE #15096: KeyName doesn't work with multi-node type autoscaler.
If user supplied ssh_private_key, currently ray autoscaler looks for KeyName in config["head_node"], but not the node type field.  Leading to a cryptic error message:
`KeyName` missing for head node.

i.e. 
```
available_node_types:
    ray.head.default:
        # The minimum number of worker nodes of this type to launch.
        # This number should be >= 0.
        min_workers: 0
        # The maximum number of worker nodes of this type to launch.
        # This takes precedence over min_workers.
        max_workers: 0
        # The node type's CPU and GPU resources are auto-detected based on AWS instance type.
        # If desired, you can override the autodetected CPU and GPU resources advertised to the autoscaler.
        # You can also set custom resources.
        # For example, to mark a node type as having 1 CPU, 1 GPU, and 5 units of a resource called "custom", set
        # resources: {"CPU": 1, "GPU": 1, "custom": 5}
        resources: {}
        # Provider-specific config for this node type, e.g. instance type. By default
        # Ray will auto-configure unspecified fields such as SubnetId and KeyName.
        # For more documentation on available fields, see:
        # http://boto3.readthedocs.io/en/latest/reference/services/ec2.html#EC2.ServiceResource.create_instances
        node_config:
            InstanceType: m5.large
            ImageId: ami-0a2363a9cff180a64 # Deep Learning AMI (Ubuntu) Version 30
            # You can provision additional disk space with a conf as follows
            BlockDeviceMappings:
                - DeviceName: /dev/sda1
                  Ebs:
                      VolumeSize: 100
            KeyName: XXXXX <- this won't work
```

```
def _configure_key_pair(config):
    if "ssh_private_key" in config["auth"]:
        _set_config_info(keypair_src="config")

        # If the key is not configured via the cloudinit
        # UserData, it should be configured via KeyName or
        # else we will risk starting a node that we cannot
        # SSH into:

        if "UserData" not in config["head_node"]:
            cli_logger.doassert(  # todo: verify schema beforehand?
                "KeyName" in config["head_node"],
                "`KeyName` missing for head node.")  # todo: err msg
            assert "KeyName" in config["head_node"]

        if "UserData" not in config["worker_nodes"]:
            cli_logger.doassert(
                "KeyName" in config["worker_nodes"],
                "`KeyName` missing for worker nodes.")  # todo: err msg
            assert "KeyName" in config["worker_nodes"]

        return config
```

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Why are these changes needed?

Closes https://github.com/ray-project/ray/issues/15096 
Addresses key_pair part of https://github.com/ray-project/ray/issues/14856

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

Added tests to check key-pair filling logic and assertion error on missing KeyName.

Did manual test: Can launch AWS cluster by specifying KeyName under available node types.

### CODE REVIEW

**Review by ijrsvt — COMMENTED** [review]
Can you modify [`log_to_cli`](https://github.com/ray-project/ray/blob/0d1b30620b27e2c65d5f51c33309e5269036efa5/python/ray/autoscaler/_private/aws/config.py#L166) to also check available node types?

File: python/ray/tests/aws/test_autoscaler_aws.py [hunk 1]
```diff
@@ -225,17 +226,38 @@ def test_fills_out_amis(iam_client_stub, ec2_client_stub):
...
+        "KeyName"] = "fakeKey"
+
+    # bootstrap_config configures iam role and then raises an assertion error
+    stubs.configure_iam_role_default(iam_client_stub)
+    with pytest.raises(AssertionError):
+        helpers.bootstrap_aws_config(config)
+
+    # expect no pending responses left in IAM client stub queue
+    iam_client_stub.assert_no_pending_responses()
+
```
Comment (ijrsvt) [inline python/ray/tests/aws/test_autoscaler_aws.py:246]: Could you add one more test to ensure the `UserData` configuration works correctly? i.e. having the `ssh_private_key` passed in and having `["node_config"]["UserData"]` set to a non-empty value

**Review by ijrsvt — COMMENTED** [review]
This overall looks good! Thanks for doing this. OOC, would this break for workflows where a user is using `available_node_types` and also setting data in `head_node` and `worker_nodes`?

**Comment by DmitriGekhtman** [comment]
> This overall looks good! Thanks for doing this. OOC, would this break for workflows where a user is using `available_node_types` and also setting data in `head_node` and `worker_nodes`?

Eh, yeah, thanks -- need to consider this carefully, as that's way it currently has to be done... In an initial version, had a check for that, then got rid of it..

**Comment by ijrsvt** [comment]
@DmitriGekhtman I think it might be good to support that workflow for a bit longer? 

What are your thoughts @AmeerHajAli

**Comment by DmitriGekhtman** [comment]
> @DmitriGekhtman I think it might be good to support that workflow for a bit longer?
> 
> What are your thoughts @AmeerHajAli

Yes, for sure. No need to change things suddenly.
Shouldn't require too many more lines of code.

File: python/ray/autoscaler/_private/aws/config.py [hunk 2]
```diff
@@ -282,17 +286,13 @@ def _configure_key_pair(config):
...
-            cli_logger.doassert(
-                "KeyName" in config["worker_nodes"],
-                "`KeyName` missing for worker nodes.")  # todo: err msg
-            assert "KeyName" in config["worker_nodes"]
+        for node_type in node_types:
+            node_config = node_types[node_type]["node_config"]
+            if "UserData" not in node_config:
+                cli_logger.doassert(  # todo: verify schema beforehand?
+                    "KeyName" in node_config,
+                    _key_assert_msg(node_type))  # todo: err msg
```
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/aws/config.py:294]: what is this todo?

File: python/ray/autoscaler/_private/aws/config.py [hunk 3]
```diff
@@ -282,17 +286,13 @@ def _configure_key_pair(config):
...
-
-        if "UserData" not in config["worker_nodes"]:
-            cli_logger.doassert(
-                "KeyName" in config["worker_nodes"],
-                "`KeyName` missing for worker nodes.")  # todo: err msg
-            assert "KeyName" in config["worker_nodes"]
+        for node_type in node_types:
+            node_config = node_types[node_type]["node_config"]
+            if "UserData" not in node_config:
+                cli_logger.doassert(  # todo: verify schema beforehand?
```
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/aws/config.py:292]: what is this todo?

File: python/ray/tests/aws/test_autoscaler_aws.py [hunk 4]
```diff
@@ -314,8 +336,15 @@ def test_temp_create_sg_multinode(iam_client_stub, ec2_client_stub):
...
-    # Confirming boostrap config does not currently touch available node types.
-    assert bootstrapped_config["available_node_types"] == config[
+    # Confirming boostrap config does not currently touch available node types,
+    # except for KeyName, which is correctly configured.
+    bootstrapped_config_copy = copy.deepcopy(bootstrapped_config)
+    node_types = bootstrapped_config_copy["available_node_types"]
+    for node_type in node_types:
+        node_config = node_types[node_type]["node_config"]
+        node_config["KeyName"] == DEFAULT_KEY_PAIR["KeyName"]
+        node_config.pop("KeyName")
```
Comment (AmeerHajAli) [inline python/ray/tests/aws/test_autoscaler_aws.py:346]: What are you trying to do?
I don't understand

File: python/ray/tests/aws/test_autoscaler_aws.py [hunk 5]
```diff
@@ -225,17 +226,38 @@ def test_fills_out_amis(iam_client_stub, ec2_client_stub):
...
+def test_missing_key(iam_client_stub):
+    """
+    Check that an error is raised when ssh_private_key is given in the
+    auth_config but a node type is missing KeyName.
+    """
+    config = helpers.load_aws_example_config_file("example-full.yaml")
+    config["auth"]["ssh_private_key"] = "~/fake/path"
+    config["available_node_types"]["ray.head.default"]["node_config"][
+        "KeyName"] = "fakeKey"
+
```
Comment (AmeerHajAli) [inline python/ray/tests/aws/test_autoscaler_aws.py:238]: Lets explicitly unset the "ray.worker_default" to make it easier to read this code and understand that there is another key that does not have a KeyName.

File: python/ray/autoscaler/_private/aws/config.py [hunk 6]
```diff
@@ -350,12 +350,23 @@ def _configure_key_pair(config):
         "Private key file {} not found for {}".format(key_path, key_name)
 
     config["auth"]["ssh_private_key"] = key_path
-    config["head_node"]["KeyName"] = key_name
-    config["worker_nodes"]["KeyName"] = key_name
+    for node_type in node_types:
+        node_config = node_types[node_type]["node_config"]
+        node_config["KeyName"] = key_name
```
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/aws/config.py:355]: When doing something like this, we should explicitly raise a warning saying that KeyName in head_node/worker_nodes are deprecated and they must be passed in the node config.

**Review by AmeerHajAli — CHANGES_REQUESTED** [review]
Can we throw a warning for using worker_nodes and head_node and mention that they are deprecated and suggest using multi node types??

**Comment by AmeerHajAli** [comment]
@DmitriGekhtman , closing this as it is no longer necessary, right?

**Comment by DmitriGekhtman** [comment]
> @DmitriGekhtman , closing this as it is no longer necessary, right?

Correct, superseded by https://github.com/ray-project/ray/pull/15584!

**Pull request closed without merging**